core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -504,7 +520,7 @@ class ProducerStateManager(
    * Truncate the producer id mapping and remove all snapshots. This resets the state of the mapping.
    */
   def truncateFullyAndStartAt(offset: Long): Unit = {
-    producers.clear()
+    removeProducerIds(producers.keySet)
```

Comment:
It's much cheaper to clear a map than to remove all the keys. It probably makes sense to have a `clearProducerIds` method as well.

Reply:
Yes, nice suggestion about `clearProducerIds`. Thanks, @ijuma .

Reply:
I asked Anastasia about this offline and she said the clear method also removes all the keys (using the -= operator for each one)

Reply:
```{ keysIterator foreach -= }```

Reply:
Not true:

```java
public void clear() {
        Node<K,V>[] tab;
        modCount++;
        if ((tab = table) != null && size > 0) {
            size = 0;
            for (int i = 0; i < tab.length; ++i)
                tab[i] = null;
        }
    }
```

Reply:
producers is currently defined as a mutable scala map. That's where I pulled the method from

Reply:
Not in trunk. :) Generally, you should not rely on a particular inefficient implementation for this exact reason.

Reply:
Justine originally did point this out to me but as it previously was a mutable scala map, we had kept it is. Since it's changed it makes sense to have a clearProducerIds method as well.